grin/test/AbstractInterpretation/EffectTrackingSpec.hs

```diff
@@ -0,0 +1,246 @@
+{-# LANGUAGE OverloadedLists, OverloadedStrings, QuasiQuotes #-}
+module AbstractInterpretation.EffectTrackingSpec where
+
+import Data.Map (Map)
+
+import Grin.TH
+import Grin.Grin
+import Grin.PrimOpsPrelude
+
+import Test.Hspec
+import Test.Assertions
+
+import AbstractInterpretation.Reduce (AbstractInterpretationResult(..),evalAbstractProgram)
+import AbstractInterpretation.EffectTracking.CodeGen hiding (live)
+import AbstractInterpretation.EffectTracking.Result
+
+runTests :: IO ()
+runTests = hspec spec
+
+calcEffects :: Exp -> ETResult
+calcEffects prog
+  | (etProgram, etMapping) <- codeGen (withPrimPrelude prog)
+  , computer <- _airComp . evalAbstractProgram $ etProgram
+  = toETResult etMapping computer
+
+spec :: Spec
+spec = describe "Effect Tracking Analysis" $ do
+
+  it "simple_print" $ do
+    let exp = [prog|
+          grinMain =
+            _prim_int_print 0
+        |]
+    let expected = mempty
+          { _register = []
+          , _function = [ ("grinMain", Effects ["_prim_int_print"]) ]
```

Comment:
Did you add tests for externals?

Reply:
The effect tracking analysis calculates only the names of the effects. It is based on externals, not primops. Are you suggesting we should collect the `External`s instead of just their names during the analysis?

Reply:
Do not rely on predefined externals in that test. Define you own external function/primop.

Reply:
Alright, i added some tests. Also added an error message to the effect tracking analysis in case the program tried to redefine an already existing external.